base/kra/src/main/java/org/dogtagpki/server/kra/rest/base/KeyRequestProcessor.java

```diff
@@ -0,0 +1,1050 @@
+    }
+
+    private KeyData createKeyData(Request request) {
+        // TODO - to be implemented when we enable one-shot generation and recovery
+        // with retrieval
+        return null;
+    }
+
+    private boolean keyExists(String clientKeyId, String keyStatus) throws EBaseException {
+
+        logger.info("KeyRequestProcessor: Checking for key existence");
+        logger.info("KeyRequestProcessor: - client key ID: {}", clientKeyId);
+        logger.info("KeyRequestProcessor: - status: {}", keyStatus);
+
+        String filter = "(" + KeyRecord.ATTR_CLIENT_ID + "=" + clientKeyId + ")";
+        if (keyStatus != null) {
+            filter = "(&" + filter + "(" + KeyRecord.ATTR_STATUS + "=" + keyStatus + "))";
+        }
+        logger.info("KeyRequestProcessor: - filter: {}", filter);
+
+        Enumeration<KeyRecord> existingKeys = repo.searchKeys(filter, 1, 10);
+
+        if (existingKeys != null && existingKeys.hasMoreElements()) {
+            logger.info("KeyRequestProcessor: Key exists");
+            return true;
+        }
+
+        logger.info("KeyRequestProcessor: Key does not exist");
+        return false;
+    }
+
+    private RequestId createEphemeralRequestID() {
+
+        JssSubsystem jssSubsystem = engine.getJSSSubsystem();
+
+        SecureRandom random = jssSubsystem.getRandomNumberGenerator();
+        long id = System.currentTimeMillis() * 10000 + random.nextInt(10000);
+
+        return new RequestId(id);
+    }
+
+    private Request createRecoveryRequest(KeyRecoveryRequest data, String userName,
+            AuthToken authToken, boolean ephemeral) throws EBaseException{
+        if (data == null) {
+            throw new BadRequestException("Invalid request.");
+        }
+
+        /*if (data.getCertificate() == null &&
+            data.getTransWrappedSessionKey() == null &&
+            data.getSessionWrappedPassphrase() != null) {
+            throw new BadRequestException("No wrapped session key.");
+        }*/
+
+        if (userName == null) {
+            throw new UnauthorizedException("Recovery must be initiated by an agent");
+        }
+
+        KeyId keyId = data.getKeyId();
+        KeyRecord rec = null;
+        try {
+            rec = repo.readKeyRecord(keyId.toBigInteger());
+        } catch (DBRecordNotFoundException e) {
+            throw new KeyNotFoundException(keyId, "key not found to recover", e);
+        }
+
+        try {
+            authz.checkRealm(rec.getRealm(), authToken, rec.getOwnerName(), "certServer.kra.key", "recover");
+        } catch (EAuthzUnknownRealm e) {
+            throw new UnauthorizedException("Invalid realm", e);
+        } catch (EBaseException e) {
+            throw new UnauthorizedException("Agent not authorized by realm", e);
+        }
+
+        RequestId requestID;
+        if (ephemeral) {
+            requestID = createEphemeralRequestID();
+        } else {
+            requestID = requestRepository.createRequestID();
+        }
```

Comment:
No issue here, I just want to point out that ephemeral request ID is probably no longer needed with RSNv3, so it's something that can be removed in the future.